Fix crash in 'sweep' and add 'sweepandsend' command to broadcast the …

todd1251 opened the pull request:
…transaction

todd1251 commented:
A summary of how this was tested:

### Generated an empty paper wallet
Address: `bNjSYPowPqTWH61ugFXVha6aH6h8Nme7rw`
Private key: `L4vCoyQR4yGi2FFn16XoR8GF5RPCGfE6x37bkTS3dbbbskTkbj4S`

### MSFTserver's paper wallet with 5 LBC (thanks! 👍)
Address: `bamwDR91pUrAFjaxv6SqtTDrtDdFs2fP1A`
Private key: `L5gQeyjJdtmLuh7PBYZyKCUsRe8BkAejsGg83XPj5sCrzquU52Ba`

![image](https://user-images.githubusercontent.com/224377/34553132-f6862ebe-f179-11e7-892b-c9a338e4480e.png)

### Restore the empty paper wallet
```
lbryum restore L4vCoyQR4yGi2FFn16XoR8GF5RPCGfE6x37bkTS3dbbbskTkbj4S
```

### Sweep MSFTserver's wallet into the empty wallet
```
lbryum sweepandsend L5gQeyjJdtmLuh7PBYZyKCUsRe8BkAejsGg83XPj5sCrzquU52Ba bNjSYPowPqTWH61ugFXVha6aH6h8Nme7rw
```
https://explorer.lbry.io/tx/e823504fdbe9a213d1f5f3f0d3989e97d35ed53075cbaef552e0dd35f1002424

### Sweep the empty paper wallet into my lbry.io client wallet
```
lbryum sweepandsend L4vCoyQR4yGi2FFn16XoR8GF5RPCGfE6x37bkTS3dbbbskTkbj4S bWVv8QGJ4bNbynCvWDquc9hyTafBTfS34p
```
https://explorer.lbry.io/tx/3414f7772136fb86d2930f197ac2868900a559a5998b1ba166de211fb3087dd1

tzarebczan commented:
Thanks @todd1251 ! Tipped you LBC on Discord last night :)

jackrobison commented on lbryum/commands.py:
```diff
@@ -471,6 +470,21 @@ def sweep(self, privkey, destination, tx_fee=None, nocheck=False):
         return Transaction.sweep(privkeys, self.network, dest, fee)
 
+    @command('n')
+    def sweep(self, privkey, destination, tx_fee=None, nocheck=False):
```
Instead of having these be separate commands, I think `sweep` should have a `broadcast` argument

kaykurokawa replied:
see here https://github.com/lbryio/lbryum/blob/master/lbryum/commands.py#L2516 for example

jackrobison commented on lbryum/commands.py:
```diff
@@ -471,6 +470,21 @@ def sweep(self, privkey, destination, tx_fee=None, nocheck=False):
+        """Sweep private keys. Broadcasts a transaction that spends UTXOs from
+        privkey to a destination address."""
+        tx = self._mksweeptx(privkey, destination, tx_fee, nocheck)
+        return self.network.synchronous_get(('blockchain.transaction.broadcast', [str(tx)]))
```
Instead of directly calling broadcast here, you should use `self.wallet.send_tx` (to avoid a race condition).

kaykurokawa replied:
see here for example: https://github.com/lbryio/lbryum/blob/master/lbryum/commands.py#L2559

kaykurokawa commented:
What was the cause of the "crash" as you mention in the original sweep command?

todd1251 commented:
@jackrobison @kaykurokawa - thanks for the review! I am working on improving the pull request per your recommendations. I arbitrarily chose to base this on `payto` / `paytoandsend`. Instead, I'll model it after one of the examples you have pointed out.

I reproduced the crash just now to find the message (below). The problem is that an attempt to reconstruct the `Transaction` object from JSON is failing because no constructor parameters are provided. This only seems to occur if the raw transaction object is returned from the command, instead of a dictionary suitable for JSON serialisation. At least, that is my understanding based on a limited review of `lbryum` and similar code in `electrum`.
```
Traceback (most recent call last):
  File "/usr/local/bin/lbryum", line 9, in <module>
    load_entry_point('lbryum==3.2.0rc8', 'console_scripts', 'lbryum')()
  File "/Users/Todd/Code/contrib/lbry/lbryum/lbryum/main.py", line 239, in main
    result = server.run_cmdline(config_options)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonrpc.py", line 288, in __call__
    return self.__send(self.__name, args)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonrpc.py", line 237, in _request
    response = self._run_request(request)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonrpc.py", line 267, in _run_request
    return_obj = loads(response)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonrpc.py", line 549, in loads
    result = jsonclass.load(result)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonclass.py", line 107, in load
    new_value = load(value)
  File "/Library/Python/2.7/site-packages/jsonrpclib/jsonclass.py", line 146, in load
    new_obj = json_class(*params)
TypeError: __init__() takes exactly 2 arguments (1 given)
```

todd1251 commented:
OK, I've implemented this using `wallet.send_tx()` but there is a problem:
```
2018-01-09 11:26:50,675 ERROR    lbryum.wallet:1248: timed out while waiting to receive back a broadcast transaction, expected txid: e2ea28074d3bfed55dcbaeae18c5699ecbed9b2312eb359fd96a5b5df472ffe6
{
  "reason": "timed out while waiting to receive back a broadcast transaction, expected txid: e2ea28074d3bfed55dcbaeae18c5699ecbed9b2312eb359fd96a5b5df472ffe6", 
  "success": false
}
```

First, the sweep does actually succeed:
https://explorer.lbry.io/tx/e2ea28074d3bfed55dcbaeae18c5699ecbed9b2312eb359fd96a5b5df472ffe6

However, because the private key is not associated with the wallet, a timeout occurs waiting for the transaction to appear in the wallet.

To avoid this, I have modified `wallet.send_tx()` to accept an optional `wait` argument. By default it is `True` and behaviour remains the same. For `sweep`, the argument is `False` and the transaction is broadcast without waiting for it to be seen in the wallet.

The output is in a similar format to the `abandon` command @kaykurokawa mentioned:
```
{
  "fee": "1.000000000000000047921736024E-12", 
  "success": true, 
  "tx": "0100000001e6ff72f45d5b6ad99f35eb12239bedcb9e69c518aeaecb5dd5fe3b4d0728eae2000000006b483045022100e4a740da2fc8abaa0f7fbac6d20159b03b08ea0fcfdea9e0f66da48e032c2e390220498e048eb1ab8ecb01cb42e916d2008e3313848027d02b8892cffaaf50e60d070121034e73e1b366e55c7d4433aef5b621c517dc1cce975456a2679531d0162177d696ffffffff01d06bf505000000001976a91498b95827c5679093bde53e941bf126642aa110b288ac00000000"
}
```

I presume this will faciliate the implementation of https://github.com/lbryio/lbry/issues/1049.

Does this solution seem reasonable, or if not, any suggestions on how to proceed?

jackrobison commented on lbryum/commands.py:
```diff
@@ -458,18 +458,29 @@ def _resolver(self, x):
+        tx = Transaction.sweep(privkeys, self.network, dest, fee)
+
+        if broadcast:
+            success, out = self.wallet.send_tx(tx, wait=False)
```
This can be `txid = self.broadcast(str(tx))`

todd1251 replied:
Great, thanks - I will take a look at use `broadcast` instead.

jackrobison commented:
>However, because the private key is not associated with the wallet, a timeout occurs waiting for the transaction to appear in the wallet.

This can use `broadcast` instead of `send_tx`, then the `wait` argument isn't necessary and it won't block on the tx being added to the history (which will only happen if the address is in the wallet and otherwise would time out).

eukreign commented on lbryum/commands.py:
```diff
@@ -458,18 +458,29 @@ def _resolver(self, x):
+                return {'success': False, 'reason': out}
+
+        return {
+            'success': (tx != None),
```
It's better to just leave the boolean operation without the parenthesis, otherwise you risk confusion:
```
'success': (tx != None),
```
looks a little too similar to:
```
'success': (tx != None,),
```
Above will actually return a Python tuple instead of just a boolean. Also, to a human reader of your code on first glance it appears that you are intending to have a Python tuple.

While,
```
'success': tx != None,
```
is very clear that the 'success' should be set to a boolean and no tuples are involved.

Only use parenthesis if it's absolutely necessary.
